core/modules/replica/DatabaseMySQL.cc

```diff
@@ -664,4 +691,110 @@ void Connection::_assertTransaction(bool inTransaction) const {
     }
 }
 
+
+ConnectionPool::Ptr ConnectionPool::create(ConnectionParams const& params,
+                                           size_t maxConnections) {
+    return Ptr(new ConnectionPool(params, maxConnections));
+}
+
+
+ConnectionPool::ConnectionPool(ConnectionParams const& params, size_t maxConnections)
+    :   _params(params),
+        _maxConnections(maxConnections) {
+}
+
+
+Connection::Ptr ConnectionPool::allocate() {
+
+    string const context = "ConnectionPool::" + string(__func__) + "  ";
+
+    LOGS(_log, LOG_LVL_DEBUG, context);
+
+    unique_lock<mutex> lock(_mtx);
+
+    // Open a new connection and return it right away if the limit hasn't
+    // been reached yet.
+    //
+    // TODO: the factory method called below may put a calling thread
+    // in the blocking state while the (database) service becomes available.
+    // This will prevent operations with the pool by other threads. Investigate
+    // a non-blocking algorithm.
+
+    if (_availableConnections.size() + _usedConnections.size() < _maxConnections) {
+        auto conn = Connection::open(_params);
+        _usedConnections.push_back(conn);
+        return conn;
+    }
+
+    // Otherwise grab an existing one (which may require to wait before
+    // it would become available).
+
+    auto const self = shared_from_this();
```

Comment:
Not sure I understand this - if you are in this method then there should be no danger that reference count drops to zero, you can pass `this` to lambda below instead?

Reply:
Good point. I have reviewed tis class, and I don't see any reason why it should be based on the `std::enable_shared_from_this` paradigm because it doesn't initiate any asynchronous operations. I'm going to simplify the class design.